docs(site): Add documentation about Ransomware Protection (#2160)
kopia/kopia: site/content/docs/Advanced/Ransomware Protection/_index.md

    * Re-generate your API master API key (to deactivate the one you just used)
    * Delete any existing Application keys that do not have restricted access

### Even more protection

So far, we have secured your access such that even if a bad actor gets access to your Kopia configuration, they can't do irreparable harm to your cloud backup.  However, what if they get access to your login credentials?  Your login credentials provide the ability to delete your data and even your entire buckets for all the buckets in your account.  But the cloud providers have protection from that too.  AWS (and S3 compatible providers) offer a feature called 'Object Locking'.  An Object Lock is applied when a file is created (or on an existing file), and it provides a specific retention date.  Until that retention date occurs, there is no way to delete the locked file.  Even using your login credentials, the file is protected from deletion.  It can still be overwritten with a new version or hidden such that it doesn't appear in a file list.  But it will always be accessible until its retention date occurs.  While Kopia supports applying Object Locks, there are some caveats:

>   * Object Locks must be enabled when a bucket is created.  If you already have backups in the cloud, you will need to create a new bucket with Object Locks turned on.

Reviewer: > You can enable Object Lock when you [create a bucket](https://www.backblaze.com/bb/docs/cloud-storage-create-and-manage-buckets) or as a one-time operation using the [b2_update_bucket](https://www.backblaze.com/apidocs/b2-update-bucket) API call. However, after you enable Object Lock on a bucket, you cannot disable this setting.

https://www.backblaze.com/docs/cloud-storage-object-lock

  * You must enable Object Lock extension in Kopia.  By default, Kopia does not renew Object Lock retention dates, however this can be enabled in the `full-maintenance` options.  You must ensure that you run full maintenance at least as frequently as your Object Lock period to ensure Object Locks do not expire, otherwise the Retention date will pass, and your data will no longer be protected by an Object Locks.
  * Kopia currently only supports object locks when using an S3 repo.  If you use Backblaze, you will need to use the S3 repo method with Kopia instead of the Backblaze native API.

### How to enable Object Locks

  * Create a new bucket with Object Locks enabled
    * There is no need to set a default Retention time
  * Optional: Create a restricted access key as instructed above for additional protection
  * Create a new Kopia repository

>     * Run: `kopia repo create s3 --bucket <bucket name> --access-key=<access key> --secret-access-key=<secret access key> --retention-mode COMPLIANCE --retention-period <retention time>`

Reviewer 2: it may be worth linking to the S3 documentation on retention modes so that folks aren't surprised by the restrictions on compliance mode

  * Object Locks must be enabled when a bucket is created.  If you already have backups in the cloud, you will need to create a new bucket with Object Locks turned on.
  * Kopia does not currently renew Object Lock retention dates.  This means that if your data doesn't change, eventually the original Retention date will pass, and your data will no longer be protected by an Object Lock even if you are taking regular snapshots.
    * However, this can be mitigated by using an external application to update the Object Locks before they expire.
  * Kopia currently only supports object locks when using an S3 repo.  If you use Backblaze, you will need to use the S3 repo method with Kopia instead of the Backblaze native API.

### How to enable Object Locks

  * Create a new bucket with Object Locks enabled
    * There is no need to set a default Retention time
  * Optional: Create a restricted access key as instructed above for additional protection
  * Create a new Kopia repository
    * Run: `kopia repo create s3 --bucket <bucket name> --access-key=<access key> --secret-access-key=<secret access key> --retention-mode COMPLIANCE --retention-period <retention time>`

>       * `retention-period`` can be specified in days (for instance 30d)

Reviewer 3: ```suggestion
      * `retention-period` can be specified in days (for instance 30d)
```

  * After each backup (or at least before the retirement period is up) it is important to reset the Object Lock retention time on all current files.
    * Kopia does not currently provide his capability, but the process is to:
      * List all filenames in the bucket (assuming a single repo per bucket or at least consistent retention times)
      * For each found file starting with 'x', 'q', or 'p': apply a PutObjectRetention command with the updated retention time.
        * files starting with _log probably do not need additional retention applied

>       * This can be achieved using a reasonably simple python script `TBD`

Reviewer 3: Update needed here WRT https://github.com/kopia/kopia/pull/2179

### How to restore a snapshot that was deleted by Ransomware (or some other process)

>   * If you have data that needs to be restored, make sure that either your retention time will not expire or your lifecycle data expiration is sufficient to ensure you can download your data before the cloud provider could remove it

Reviewer 3: ```suggestion
  * If you have data that needs to be restored, make sure that either your retention time will not expire or your lifecycle data expiration is sufficient to ensure you can download your data before the cloud provider removes it
```

  * Disconnect the repo in Kopia

>   * Reconnect the repo in Kopia using the `--point-in-time` option (ex: --point-in-time=2021-11-29T01:10:00.000Z`)

Reviewer 3: ```suggestion
  * Reconnect the repo in Kopia using the `--point-in-time` option (ex: `--point-in-time=2021-11-29T01:10:00.000Z`)
```

    * This option is only currently available when using an 's3' repo

### A caveat about Ransomware protection

Ransomware can be very sophisticated software.  It may run on your system for weeks, slowly encrypting data that is not frequently used, and only hitting actively used data last (to avoid detection for as long as possible).  This means that you should ensure your data retention period is long enough that you will be able to recover your data.  It also means that recovery may not be easy.  You may have many snapshots after the Ransomware has started modifying your system, and you may have made changes throughout that time such that there is no single snapshot that represents a good state of all of your files.  Restoration is likely to be a time-consuming task, and your best bet is to protect your system from these threats so they don't occur in the first place.  But if they do, Kopia can provide some additional security to protect your data.

> Additionally note that ransomware could theoretically weaponize object-locks to cost you a lot of money.  Becuase object-locks cannot be shortened are removed, Sufficiently malicious ransomware could upload large amounts of data and set a very long object lock that would make it impossible to delete.  It is strongly recommended to ensure you have appropriate quotas/limits on your buckets to limit potential storage costs.

Reviewer 3: There seems to be an inconsistency with capitalization of the word 'ransomware' in this doc: sometimes it's capitalized, sometimes not. I suggest we change all instances (not at the beginning of a sentence of course) to `ransomware`, consistent with what I'm seeing on wikipedia and news articles.

> Additionally note that ransomware could theoretically weaponize object-locks to cost you a lot of money.  Becuase object-locks cannot be shortened are removed, Sufficiently malicious ransomware could upload large amounts of data and set a very long object lock that would make it impossible to delete.  It is strongly recommended to ensure you have appropriate quotas/limits on your buckets to limit potential storage costs.

Reviewer 3: ...`cannot be shortened are removed` --> `cannot be shortened or removed`

`removed, Sufficiently malicious` --> `removed, sufficiently malicious`
